lib/jsonrpc_client.ml

```diff
@@ -17,49 +17,95 @@
 module D = Debug.Make(struct let name = "jsonrpc_client" end)
 open D
 
-let input_json_object fin =
-	let buf = Buffer.create 1024 in
-	let brace_cnt = ref 0 in
-	let in_string = ref false in
-	let last_char () = Buffer.nth buf (Buffer.length buf - 1) in
-	let rec get () =
-		let c = input_char fin in
+exception Timeout
+exception Read_error
+
+let json_rpc_max_len = ref 65536 (* Arbitrary maximum length of RPC response *)
+let json_rpc_read_timeout = ref "60000000000"
+let json_rpc_write_timeout = ref "60000000000"
+
+let to_s  s = (Int64.to_float s) *. 1e-9
+
+(* Read the entire contents of the fd, of unknown length *)
+let timeout_read fd timeout =
+	let buf = Buffer.create !json_rpc_max_len in
+	let read_start = Mtime_clock.counter () in
+	let get_total_used_time () = Mtime.Span.to_uint64_ns (Mtime_clock.count read_start) in
+	let rec inner max_time max_bytes =
+		let (ready_to_read, _, _) = Unix.select [fd] [] [] (to_s max_time) in
+		(* This is not accurate the calculate time just for the select part. However, we
+		 * think the read time will be minor comparing to the scale of tens of seconds.
+		 * the current style will be much concise in code. *)
+		let used_time = get_total_used_time () in
+		let remain_time = Int64.sub timeout used_time in
+		if remain_time < 0L then 
 		begin
-			match c with
-			| '{' when not !in_string -> brace_cnt := !brace_cnt + 1
-			| '}' when not !in_string -> brace_cnt := !brace_cnt - 1
-			| '"' when !in_string && (last_char () <> '\\') -> in_string := false
-			| '"' when not !in_string -> in_string := true
-			| _ -> ()
+			debug "Timeout after read %d" (Buffer.length buf);
+			raise Timeout
 		end;
-		Buffer.add_char buf c;
-		if !brace_cnt > 0 then
-			get ()
+		if List.mem fd ready_to_read
+		then
+		begin
+			let bytes = Bytes.make 4096 '\000' in
+			match Unix.read fd bytes 0 4096 with
+				| 0 -> Buffer.contents buf (* EOF *)
+				| n ->
+					if n > max_bytes
+					then
+					begin
+						debug "exceeding maximum read limit %d, clear buffer" !json_rpc_max_len;
+						Buffer.clear buf;
+						raise Read_error
+					end
+					else
+					begin
+						Buffer.add_subbytes buf bytes 0 n;
+						inner remain_time (max_bytes - n)
+					end
+				| exception Unix.Unix_error(err,_,_) when err = Unix.EAGAIN || err = Unix.EWOULDBLOCK ->
+					inner remain_time max_bytes
+		end
+		else inner remain_time max_bytes
```

Comment:
We are already timeout here, why not raise `Timout`?

Reply:
Why? Couldn't it be that something else triggered the select?

Reply:
We are only watching one fd, will the `select` return when other fds which we are not watching is ready to read?

Reply:
Yes, you are right. We could do `if remain_time < 0L or not (List.mem fd ready_to_read) then ...` and get rid of the second `if` altogether.

Reply:
Exactly, but do we need to consider the third element of the return value of `select` which contains the fds that exception occurs?

Reply:
Does it return something even if we are not watching it for exceptional cases?

Reply:
I don't think so, so we don't need to care about that.

Reply:
@krizex , I cannot agree here. It is how and when select will return is not accurately defined by Ocaml lib, so I think the current code is necessary. Usually in c coding with native OS select, it is necessary to calculate the real time after returned, as select could prematurely return as interrupted by other signal or events from the IO. I doubt Ocaml lib will wrap something to make the unix.select more 'robust' to such event, as usually such event is practically necessary in some situation.

Reply:
So you should catch the `Unix_error` with argument `EINTR` from `select` with an exception handler. And `Unix.read` and `Unix.single_write` as well just in case. Remember the discussion about `really_write` from a while ago?

Reply:
I think Rob is right, refer to https://github.com/ocaml/ocaml/blob/13945a71ed2616df9a4672cfc83f8f450ec437a8/otherlibs/unix/select.c#L96